### PR TITLE
Make `@intFromEnum` an error for empty enums

### LINKED ISSUE #18767: compiler crashes (`@intFromEnum`?)
hi, i wanted to reimplement wayland in zig (mostly to learn wayland and zig, maybe use it in an app). i really liked the idea of subsurfaces, and i wanted to make them be able to move across the screen, so i started implementing a function that takes two lists of subsurfaces and writes move and resize requests only when needed. i started writing that in tree.zig, and now the compiler started segfaulting

### Zig Version

0.11.0 (flatpak)

### Steps to Reproduce and Observed Behavior

```
git clone https://codeberg.org/two/zdg
cd zdg
./makewp.sh
zig build-obj tree.zig
```

### PR DESCRIPTION
**Pull request by Rexicon226**
fixes #18767

+ makes running `@intFromEnum` on undefined return undefined
+ makes running `@intFromEnum` on empty enums (that are exhaustive) an error.

~~should we reconsider empty enums in general? they have no actual value representation, and so i'm confused why we have them.~~

### CODE REVIEW

**Comment by rohlem** [comment]
> should we reconsider empty enums in general? [...] i'm confused why we have them.

Empty `enum`-s are uninstantiable ("`noreturn`-like") types - see https://github.com/ziglang/zig/issues/15909 .
Short version: Using such a type somewhere asserts to the compiler that it will never be instantiated.
For example if someone reifies an `enum` via `@Type` with conditional fields, if none of the fields end up being present, then a union field having this type may never be active - it would be a logic bug (=> compile error) if the code still tried to access, since there is no possible value to read/write to it.

**Review by Vexu — CHANGES_REQUESTED** [review]
> makes running @intFromEnum on undefined an error, which is something that has bitten me too many times.

It should return `undefined`.

>should we reconsider empty enums in general? they have no actual value representation, and so i'm confused why we have them.

See the issues I linked in my comment.

**Comment by mlugg** [comment]
This PR makes one change which is objectively wrong (erroring on `undefined`) and another which depends on undecided language semantics (#15909 and friends). I don't think there would be any value in merging it.

**Comment by Rexicon226** [comment]
I hear the people, and undefined it shall return.

However in response to @mlugg I disagree. This is a patch for the issue. I think that I should add a link and a TODO to it. see tagName for the exact same thing. We should not keep a crash for an undefined (no pun) amount of time because of a non resolved issue. I feel the solution is very clear, but I understand it's under debate.

File: src/Sema.zig [hunk 1]
```diff
@@ -8599,8 +8600,22 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
...
+    // https://github.com/ziglang/zig/issues/15909
+    if (enum_tag_ty.enumFieldCount(mod) == 0 and !enum_tag_ty.isNonexhaustiveEnum(mod)) {
+        return sema.fail(block, operand_src, "cannot use @intFromEnum on empty enum '{}'", .{
+            enum_tag_ty.fmt(mod),
+        });
+    }
+
     const int_tag_ty = enum_tag_ty.intTagType(mod);
 
+    if (try sema.resolveValue(operand)) |resolved_op| {
```
Comment (Vexu) [inline src/Sema.zig:8613]: Replace the other use of `resolveValue` with `resolveDefinedValue` instead.

Comment (Rexicon226) [reply src/Sema.zig:8613]: Is this not exactly what you were telling me not to do?
```zig
if (val.isUndef(mod)) {
    return sema.failWithUseOfUndef(block, src);
}
```

Comment (Vexu) [reply src/Sema.zig:8613]: No, I was wrong. But you should still put that `isUndef` in the other `resolveValue` branch.

Comment (Rexicon226) [reply src/Sema.zig:8613]:
```zig
if (try sema.resolveValue(enum_tag)) |enum_tag_val| {
    if (enum_tag_val.isUndef(mod)) {
        return mod.undefRef(int_tag_ty);
    }

    const val = try enum_tag_val.intFromEnum(enum_tag_ty, mod);
    return Air.internedToRef(val.toIntern());
}
```?

Comment (Vexu) [reply src/Sema.zig:8613]: Yes something like that, or you could make `Value.intFromEnum` handle `.undef`.

Either way there is still some other bug that I don't have time to look into now but it's in the comment I made in the issue.

Comment (Rexicon226) [reply src/Sema.zig:8613]: I pushed this, so resolving.

**Pull request merged**